presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java

```diff
@@ -1567,6 +1584,89 @@ public PhysicalOperation visitJoin(JoinNode node, LocalExecutionPlanContext cont
             }
         }
 
+        private PhysicalOperation createSpatialJoin(JoinNode node, LocalExecutionPlanContext context)
+        {
+            Verify.verify(node.getFilter().isPresent());
+            Expression filterExpression = node.getFilter().get();
+            List<FunctionCall> spatialFunctions = extractSupportedSpatialFunctions(filterExpression);
+            for (FunctionCall spatialFunction : spatialFunctions) {
+                List<Expression> arguments = spatialFunction.getArguments();
+                Verify.verify(arguments.size() == 2);
+
+                if (!(arguments.get(0) instanceof SymbolReference) || !(arguments.get(1) instanceof SymbolReference)) {
+                    continue;
+                }
+
+                String firstSymbol = ((SymbolReference) arguments.get(0)).getName();
+                String secondSymbol = ((SymbolReference) arguments.get(1)).getName();
+
+                PlanNode probeNode = node.getLeft();
+                Set<String> probeSymbols = getSymbolNames(probeNode.getOutputSymbols());
+
+                PlanNode buildNode = node.getRight();
+                Set<String> buildSymbols = getSymbolNames(buildNode.getOutputSymbols());
+
+                if (probeSymbols.contains(firstSymbol) && buildSymbols.contains(secondSymbol)) {
+                    return createSpatialLookupJoin(node,
```

Comment:
you cannot return so early, what if there are more spatial functions in this join?

Reply:
I feel that optimizing spatial joins with multiple spatial relationships is outside of the scope of this PR.